### PR TITLE
Initial support for IRC v3.2 tags

### LINKED ISSUE #217: Make IRC Capabilities configurable
I was getting in some struggle while connecting to a IRCv3 server where i needed to request custom IRC capabilities.

For this I worked something out:

`cinch/configuration/bot.rb: Line 14`

``` ruby
  def self.default_config
    {
      :server => "localhost",
      :port   => 6667,
      # ...
      :caps => []
    }
```

`cinch/irc.rb: Line 114`

``` ruby
    def send_cap_req
      # Getting capabilities from bot config
      caps = @bot.config.caps & @network.capabilities
      if caps.size > 0
        send "CAP REQ :" + caps.join(" ")
      else
        send_cap_end
      end
    end
```

Example:

``` ruby
bot = Cinch::Bot.new do
    configure do |c|
        c.server    = "irc.example.com"
        c.password  = "password"
        c.channels  = ["#channel"]
        c.nick      = "nick"
        c.verbose   = true
        # awesome custom caps
        c.caps      = [:"example.com/customcap1", :"example.com/customcap2"]
    end
    # ...
end
```

### PR DESCRIPTION
**Pull request by Defman21**
Fixes #217 #201

### CODE REVIEW

**Comment by Defman21** [comment]
Note: the regex has been changed to provide compatibility with older versions of IRC (which doesn't have tags supported). I've tested it using Twitches IRC PRIVMSG without tags enabled (which makes Twitch IRC responds without IRC3.2 tags) and it works fine.

**Comment by Defman21** [comment]
Also I'm wondering what will be better: `#<Message:>.tags[:property]` or `#<Message:>.tags.property` (requires OpenStruct).

File: cinch.gemspec [hunk 1]
```diff
@@ -1,6 +1,6 @@
 Gem::Specification.new do |s|
   s.name = 'cinch'
-  s.version = '2.3.1'
```
Comment (txdv) [inline cinch.gemspec]: The versions shouldn't be incremented.

Comment (Defman21) [inline cinch.gemspec]: Gonna fix

Comment (Defman21) [inline cinch.gemspec]: Fixed

**Comment by dominikh** [comment]
Hey, thanks for the PR. I'm currently swamped with other work, but I'll try to take a look at it soon.

**Comment by bougyman** [comment]
The regex ```+      match = @raw.match(/(?:^@([^:]+))?(?::?(\S+) )?(\S+)(.*)/)` will not work the way you expect. I think we need some code logic here and not try to stuff everything into one regex. If there's no first capture, @tags will get the 2nd capture, etc.

**Comment by Defman21** [comment]
If there's no first capture, `@tags` will be `nil`.

```ruby
match = "bye world".match(%r/(hello)? (world)/) # => #<MatchData " world" 1:nil 2:"world">
hello, world = match.captures # => [nil, "world"]
```

**Comment by bougyman** [comment]
Thanks for the info. I've been bitten by this before, but your method looks safe from this.

**Pull request merged**